Feature/CSL-2046 v2 student detail page

klown opened the pull request:
Branch for tracking all the student details page work, including its sub-tasks.

JIRA: https://castudl.atlassian.net/browse/CSL-2046

mbrambilla committed:
CSL-2100: Student details page and reaction modal mockups

Also fixes typo in scss: divder -> divider

mbrambilla committed:
CSL-2100: Remove time view from table container

klown committed:
CSL-2046: Add skeleton template for the student details page

klown committed:
CSL-2046: Add the actual student details template file

klown committed:
CSL-2046: Add student links in the Dashboard's activity panel

Each student listed in the "Active reading time" panel is now a link
to the Student detail page.

klown committed:
CSL-2046/CSL-2106: Implement the student selector

Make the "Student Details" popup functional.

klown committed:
CSL-2046: Replace some mockup markup with template variables

klown committed:
CSL-2046: Use a `clusive_user` QuerySet for period roster; fix bug

The bug was that if the "student" name in the request url was the
teacher of the class, it would show up at the top of the "Student
Details" popup.  The fix was changing the filter for the QuerySet
to include only students and guests.  And, if a non-student is
passed in the request, it is treated as a student that is not in
the period.

cindyli committed:
CSL-2102: Implement "activity" box on the student details page

cindyli committed:
CSL-2102: fix a bug to display activities for the selected clusive user

klown committed:
CSL-2015: View/template interaction so the affect wheel displays properly

Note: using fake affect data.

cindyli committed:
CSL-2107: Implement time scale sele selector on the student details page

klown committed:
CSL-2015: Include a partial template for the Reactions panel

cindyli committed:
Merge branch 'feature/CSL-2107-student-details-time-scale-selector' into feature/CSL-2102-student-details-activities-panel

cindyli committed:
CSL-2102: Activity panel now reflects the selection on the time scale

cindyli committed:
CSL-2107: Remove the support of URL 'detailss/<username>'

klown committed:
Merge branch 'development' into feature/CSL-2046-V2-student-detail-page

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2107-student-details-time-scale-selector

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2102-student-details-activity-panel

cindyli committed:
CSL-2102: Move the shared code out of the code for the activity panel

klown committed:
CSL-2105: Add "all" timeframe for reaction data for the student

klown committed:
CSL-2105: Show modal affect details for the current student, not the class

When clicking one of the affect words in the "Reactions" panel show the
current student's reactions modal, not the request's `clusive_user`.
Since the latter is a teacher, that would pop up the class reactions modal.

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

cindyli committed:
CSL-2103: Implement the topics panel on the student details page

klown committed:
CSL-2105: Attempt at using Events to get the affect details (incomplete)

Uses Events for the student/period in the given time frame that were
either `affect_check_words` or `affect_check_free_response` as their
`control` value.

klown committed:
CSL-2105: Create temporary AffectiveUserTotal instance for the affect wheel

Firxt, retrieve the AffectiveCheckResponse objects from the database for
the student in the given time frame, then create an AffectiveUserTotal and
load it with the values from the AffectiveCheckResponse objects, but DO NOT
save the just created AffectiveUserTotal instance to the database.

klown committed:
CSL-2051: Add modal for student details page reactions for a single student

- modal shows the book titles and the affect free responses to those
  titles.
- wire frame: https://castudl.atlassian.net/browse/CSL-1847?focusedCommentId=36931
- TO DO: fix the affect check responses retrieved to be based on the
  student details page time frame.  It currently uses the
  `AffectDetailView.recent_with_word()` class method and uses the
  five most recently modified.

cindyli committed:
CSL-2107: Save days selected on the student details page into db

cindyli committed:
Merge branch 'feature/CSL-2107-student-details-time-scale-selector' into feature/CSL-2102-student-details-activity-panel

cindyli committed:
Merge branch 'feature/CSL-2102-student-details-activity-panel' into feature/CSL-2103-student-details-topics-panel

klown committed:
CSL-2051: Use current time scale to populate the modal invoked from affect links

Given the current time scale for the student details page, use it when
retrieving the most recent affect check responses from the database.

bgoldowsky committed:
Merge pull request #349 from cindyli/feature/CSL-2107-student-details-time-scale-selector

CSL-2107: Implement time scale selector on the student details page

klown committed:
Merge remote-tracking branch 'origin/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

klown committed:
CSL-2105: Add better messaging re: time scale for affect modal

klown committed:
CSL-2105: Add better messaging re: time scale for Reactions panel

klown committed:
CSL-2105: Consistent spacing

cindyli committed:
CSL-2102: address code review suggestions

klown committed:
CSL-2105: Clean out some unused variables and add one optimization check

klown committed:
CSL-2105: Another minor database optimization

Use count() instead of len().

cindyli committed:
CSL-2102: address more code review comments

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2102-student-details-activity-panel

cindyli committed:
CSL-2102: Swap method names for getting reading data

klown committed:
Merge pull request #350 from cindyli/feature/CSL-2102-student-details-activity-panel

CSL-2102: Implement the activity panel on the student details page

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2103-student-details-topics-panel

cindyli committed:
CSL-2103: Update variable names due to the merge

cindyli committed:
CSL-2103: Add error handling when the student isn't in the current period

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

klown committed:
CSL-2046: Refactor to put check for `None` student in one place

When the student in the request is not in this class/period, set the
various panel data values to "None", zero, or whatever is appropriate
for that panel to indicate that there is no data for it.  This is done
once in an `except` clause rather than checking for a `None` student
for each panel.

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

klown committed:
CSL-2105: Moved code around due to previous merge with upstream feature/CSL-2046-V2-student-detail-page branch

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2103-student-details-topics-panel

klown committed:
Merge branch 'development' into feature/CSL-2046-V2-student-detail-page

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

klown committed:
CSL-2105: Compromise to show all students reaction for the `None` student case

The idea is that if the student in the request is not in the period, show
the class reactions so that the teacher can look at that summary and decide
which student's details they want to investigate.

klown committed:
CSL-2102: Use ParadataDaily to track words looked up on a daily basis

klown committed:
CSL-2102: Implement "Words looked up" panel on the student details page

mbrambilla committed:
CSL-2100: Active sort indicator

klown committed:
Merge branch 'feature/CSL-2100-student-details-mockup' into feature/CSL-2046-V2-student-detail-page

klown committed:
CSL-2046: Removed "export" link from reading details table

cindyli committed:
CSL-2101: Populate initial reading details data

klown committed:
CSL-2046: Some differences from the mockup, hand copied into the template

mbrambilla commented on src/roster/templates/roster/student-details.html:
```diff
@@ -0,0 +1,203 @@
+        </div>
+    </div>
+</main>
+{# Unsure if the tour or tips are available on the student detail page #}
+{# Leaving in place for now #}
+{% for tour in tours %}
+{% include "shared/partial/popover_tour.html" with tour_name=tour.name tour_robust=tour.robust %}
+{% endfor %}
+{% if tip_name %}
+{% include "shared/partial/tooltip_tip.html" %}
+{% endif %}
```
Not needed as this is already included in the [base.html](https://github.com/cast-org/clusive/blob/40837425056fcf1f65420ddb4067280031730b76/src/shared/templates/shared/base.html#L74-L79) template.  This template builds on top of the base.

klown replied:
Thanks @mbrambilla ; removed.  But, I've still not added the `libraryMasonryEnable();` script until the layout issue with the template and FF is worked out.

@cindyli this is now ready to merge with your pull requests to get the correct styles.

mbrambilla replied:
@klown Just double checked with MacOS and Win FF with the function call - the Masonry layout is working fine for me.  Chrome, Safari, and Edge are also working.  Checked with reduced motion/animation OS settings also.

Could it be a browser cache issue?

klown replied:
@mbrambilla  Just now I added the script to the template for the first time today, and the layout of the details page is fine in FF.  I've pushed the change here.

@cindyli The student details template has one more change.  Could you merge it with your work and see if the layout is okay for you, too?  Thanks.

cindyli replied:
@klown Merged this branch into my working branch. Tested with FF, Chrome and Safari on Mac. The layout looks fine on them.

klown replied:
🎊

klown committed:
CSL-2046: More small changes to the template to match the CSL-2100 mockup

TODO: add `libraryMasonryEnable();` script for top panels layout.

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

cindyli committed:
CSL-2101: Optimize db queries for Paradata

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2102-ParadataDaily-track-words-looked-up

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2101-reading-details

klown committed:
CSL-2104: Batch load of ParadataDaily words looked up: Partial implementation

cindyli committed:
CSL-2101: implement sorting function for reading details panel

klown committed:
CSL-2046: Add `libraryMasonryEnable()` for the layout of the upper panels

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2101-reading-details

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2103-student-details-topics-panel

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2102-ParadataDaily-track-words-looked-up

cindyli committed:
CSL-2101: Save the sorting column for reading details into database

cindyli committed:
CSL-2101: Implement paging control for reading details

cindyli committed:
CSL-2101: Adjust paging data for reading details

klown committed:
CSL-2104: Complete batch load of ParadataDaily field `words_looked_up`

cindyli committed:
CSL-2101: Polish the display of words looked up

bgoldowsky commented on src/roster/views.py:
```diff
@@ -1466,3 +1467,64 @@ def remove_social_account(request, *args, **kwargs):
+            self.days = self.clusive_user.student_activity_days
+
+        period = self.clusive_user.current_period
+        self.roster = period.users.exclude(user=request.user, role=Roles.TEACHER).order_by('user__first_name')
```
This should look for STUDENT role.  Currently it incorrectly includes parents in the displayed info.

klown replied:
Good catch.  I should add more parent users since, as it is now, I won't see any parents when I test this.

bgoldowsky commented on src/roster/views.py:
```diff
@@ -1466,3 +1467,64 @@ def remove_social_account(request, *args, **kwargs):
+            self.clusive_student = ClusiveUser.objects.get(
+                user__username=kwargs['username'],
+                periods__in=[period],
+                role__in=[Roles.STUDENT, Roles.GUEST]
```
Guest not needed - would never be a member of a period.

klown replied:
Makes sense.

klown replied:
Note to self, change to:
```
self.clusive_student = ClusiveUser.objects.get(
    user__username=kwargs['username'],
    periods__in=[period],
    role=Roles.STUDENT
)
```

bgoldowsky reviewed (commented):
src/roster/templates/roster/student-details.html should be renamed with an underscore instead of a hyphen, to match all the other templates.

klown committed:
CSL-2104: Exclude word lookup events that do not reference a book

klown committed:
CSL-2104: Add management command to check the word lookup lists

This compares the `words_looked_up` fields of a group of ParadataDaily
records with their parent Paradata.  It combines the looked-up words for
all ParadataDaily records into a single set and then compares that union
with the parent Paradata's `words_looked_up`.  They should match.  If they
do not match, an error message is printed.

klown committed:
CSL-2104: Remove improper model import from word lookup sanity check script

cindyli committed:
CSL-2101: Address code review suggestions

klown commented:
> src/roster/templates/roster/student-details.html should be renamed with an underscore instead of a hyphen, to match all the other templates.

It looks like I copied the name from the mockup:  `src/frontend/html/student-details.html`

I'll use `git mv` to change the name.

klown committed:
CSL-2046: Address code review suggestions

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2103-student-details-topics-panel

klown committed:
Merge pull request #353 from cindyli/feature/CSL-2103-student-details-topics-panel

CSL-2103: Implement the topics panel on the student details page

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2101-reading-details

klown committed:
CSL-2046: Fix reference to renamed `student_details.html` template

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2101-reading-details

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2102-ParadataDaily-track-words-looked-up

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

klown committed:
CSL-2105: Rename a template file and update a reference to it

klown committed:
CSL-2105: Address review comments

klown committed:
CSL-2046: When getting the roster, match against student role

Before this change, the match was against a list of roles.  At one time
this list included the guest role. Even after the guest role was
removed, the match remained against a list now containing just the
student role.  This change removes the list match entirely.

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2102-ParadataDaily-track-words-looked-up

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

cindyli committed:
CSL-2101: remove a debugging line

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2101-reading-details

cindyli committed:
CSL-2101: use f-string to concatenate messages

bgoldowsky committed:
Fix links and avoid extra query on detail modal.

klown committed:
Merge pull request #356 from cast-org/feature/CSL-2102-ParadataDaily-track-words-looked-up

CSL-2104 Words looked up panel using ParadataDaily

cindyli committed:
Merge remote-tracking branch 'upstream/feature/CSL-2046-V2-student-detail-page' into feature/CSL-2101-reading-details

klown committed:
Merge pull request #358 from cindyli/feature/CSL-2101-reading-details

CSL-2101: Implement the reading details panel on the student details page

klown committed:
Merge branch 'feature/CSL-2046-V2-student-detail-page' into feature/CSL-2105-reactions-panel-V2-student-details

klown committed:
Merge pull request #351 from cast-org/feature/CSL-2105-reactions-panel-V2-student-details

CSL-2105 "Reactions" panel for v2 student details

klown committed:
Merge branch 'development' into feature/CSL-2046-V2-student-detail-page

bgoldowsky commented:
I have more testing to do but a couple of comments from a little poking around:
* We should think about empty state(s).
* If there is only one version of a reading, should not show "Challenge 1" in the last column.
* A line break between the hours and visits in the second column would improve the layout.
* Open responses to the standard comp question are not showing up in the "learning" column, just the nothing/a little/a lot rating. Are they supposed to?
* If a book is customized but without any custom question, "()" shows up instead of "—" in the 6th column.

klown commented:
Aside, from the empty states issue, these are all relative to the reading details table/panel.  I'll start working on them.
> 
> * If there is only one version of a reading, should not show "Challenge 1" in the last column.
> * A line break between the hours and visits in the second column would improve the layout.
> * Open responses to the standard comp question are not showing up in the "learning" column, just the nothing/a little/a lot rating. Are they supposed to?
> * If a book is customized but without any custom question, "()" shows up instead of "—" in the 6th column.

In addition, I've noticed that in one case some of the links in the student affect wheel lead to the modal saying, for example, "No readings have made Samuel feel okay yet." even though the "okay" slice is filled in. I'll look into that too.

klown committed:
CSL-2046: Add line break between hours and visits for contents of the "Active time" column

klown committed:
CSL-2046: Modify "Topics" and "Reading details" panels

- Topics:
  - Show "None" if the student has not read any books in the
    current time period,
- Reading details:
  - In the "Learning" column, show both the scaled value and the
    student's free response, if any,
  - Show "No readings in the past <time period>" if the student
    did not read anything in the current time period.

klown committed:
CSL-2046: Modify "Reading details" panel's "Reading Level" column

Do not display  challenge level when the reading has only one version.

klown committed:
CSL-2046: Fix display of the "Custom" column in the reading details panel

Refactored to handle cases where one or both of the custom question
and the student's response are None or the empty string.

klown committed:
CSL-2046: Fix date display in the "Last read" column of the Reading details panel

klown commented:
@bgoldowsky Your issues as well as the way to display the date for the "Last read" column have been addressed:

- Empty states handled as suggested by [Kim's comment](https://castudl.atlassian.net/browse/CSL-2046?focusedCommentId=37671) and [your response](https://castudl.atlassian.net/browse/CSL-2046?focusedCommentId=37676) in the JIRA,
- "Challenge" removed from "Reading level" column when there is only one version of the reading,
- Line break added between hours and visits in the "Active time" column,
- Free responses combined with ratings in the "Learning" column when a response is available,
- Show "—" in "Custom" column when there is no custom question.
- Dates in the "Last read" column displayed as "Day, Mon nn", e.g., "Fri, Nov 25" as per [Agatha's recommendation](https://castudl.atlassian.net/browse/CSL-2046?focusedCommentId=37697).

@mbrambilla regarding the display of last read date, I added a `style=white-space:nowrap';` to the `<td>` element to keep the date from wrapping.  I looked through the css files in the `frontend` folder, but I couldn't find a class that used that style without a lot of other styles.  Is there one?  Here's a reference to the use in the student's reading details table -- thanks for any advice here:
 https://github.com/cast-org/clusive/blob/37459b760655770d9c7e2c77608e19c6048e0ddc/src/roster/templates/roster/partial/student_details_reading_details.html#L70

klown commented:
I wrote:

> In addition, I've noticed that in one case some of the links in the student affect wheel lead to the modal saying, for example, "No readings have made Samuel feel okay yet." even though the "okay" slice is filled in. I'll look into that too.

I traced this back to the database and found that (1) there are `AffectiveUserTotal` records for the student with counts for certain affect words, but (2) no corresponding `AffectCheckResponse` records for that word.  Since the dialog that is presented when clicking on the reaction wheel is based on the `AffectCheckResponse` records found, it reports that no readings made the user react that way.

Since my local development database is not pristine, I suspect it's a case of GIGO, and that there is no real bug here.

klown commented:
I had another thought about `AffectiveUserTotal`s and `AffectCheckResponse`s.  The latter refer to a book, and if that book is deleted, the check response is also deleted.  The `AffectiveUserTotal`s are not affected, however, since they have no direct connection to any readings.  So, it is possible for a user's affect summaries to hold reactions but lose the context (the readings) of that reaction.

bgoldowsky commented:
> * "Challenge" removed from "Reading level" column when there is only one version of the reading,

I am not seeing challenge listed at all now, even when there are multiple versions.

bgoldowsky commented:
OK, so other than the small issue mentioned above, this is working very well. I do think we should do a round of efficiency improvements, however. On a modest sample user (who has about 15 readings), DJDT is showing 259 SQL queries were run.  A couple of thoughts on how to cut this down:

1. Use `prefetch_related` on some queries so that information is already present without having to do separate queries later. For example in `get_reading_data`, it queries for all the user's Paradatas, then there are separate queries that look up the Book for each Paradata and whether it was assigned.
2. Only call `get_book_details` for the 10 books that are going to be displayed on the current page, not all books they ever read.

klown committed:
CSL-2046: Fix bug to show "Challenge" levels

klown commented:
@bgoldowsky wrote:
> I am not seeing challenge listed at all now, even when there are multiple versions.

The latest push should fix that.  I also added a line break between the reading levels and the challenge level, as shown in the wire-frames.

I'll start looking at the db access improvements you suggested.

klown committed:
CSL-2046: Fetch only the details for the readings displayed on the page.

klown commented:
@bgoldowsky, the latest push calls `get_book_details()` for just the books shown on the page.  It's noticeably faster when changing the sort and when paging to a given page.

However, there is some fiddling so that the paging control works correctly that I'm less satisfied with.  It's good enough for now.  I'll next look into `prefetch_related` next, but if I think of improvements for paging, I'll switch back to that.

klown committed:
CSL-2046: Prefetch book objects when retrieving students' paradata records

klown committed:
CSL-2046: Move pagination constants to the settings file

klown committed:
Merge branch 'development' into feature/CSL-2046-V2-student-detail-page

klown committed:
CSL-2046: Remove errant 'import pdb'

klown committed:
CSL-2046: Another `prefetch_related()` database optimization

bgoldowsky committed:
CSL-2046: Consistency of date span + small tweaks.
Made it consistent that data includes the date 7 days ago; was gte for the panels vs. gt for the list display.
A couple of tiny efficiency/style improvements.
